### PR TITLE
Fixes Local MD implementation

### PR DESCRIPTION
**Pull request by badisa**
This PR is to get local MD to a state where it is correct. There were several bugs in the original merge of the implementation. With these changes local md no longer has any performance improvement over global md. Changes to make it faster will come in future PRs.

### Bugfixes
* Use the restraints
* Ran NonbondedAllPairs on subset of system which did not consider particles moving beyond its starting point.
* Fixed the density test to properly fail (https://gitlab.com/relaytx/physics-platform/timemachine/-/pipelines/797870948)

### Additions
* Initialize/finalize integrators
* Add a benchmark (performance expected to be similar if not slower then global MD)

### Benchmark
From https://gitlab.com/relaytx/physics-platform/timemachine/-/jobs/3884085148

```
hif2a-rbfe: N=8840 speed: 588.37ns/day dt: 1.5fs (ran 200 steps in 0.04s)
hif2a-rbfe-local: N=8840 speed: 548.95ns/day dt: 1.5fs (ran 200 steps in 0.05s)
```

### CODE REVIEW

File: timemachine/cpp/src/context.cu [hunk 1]
```diff
@@ -332,12 +332,15 @@ std::array<std::vector<double>, 2> Context::multiple_steps_local(
...
+        // particles that aren't included in frozen
+        // set_nonbonded_potential_idxs(nonbonded_potential, p_num_selected.data[0], d_row_idxs.data, stream);
+        intg_->initialize(local_bps, d_x_t_, d_v_t_, d_box_t_, d_shell_idxs_inner.data, stream);
         for (int i = 0; i < burn_in; i++) {
-            this->_step(bps_, d_shell_idxs_inner.data, stream);
+            this->_step(local_bps, d_shell_idxs_inner.data, stream);
         }
         for (int i = 1; i <= n_steps; i++) {
-            this->_step(bps_, d_shell_idxs_inner.data, stream);
+            this->_step(local_bps, d_shell_idxs_inner.data, stream);
```
Comment (maxentile) [inline timemachine/cpp/src/context.cu:343]: `bps_` -> `local_bps` is the primary bug-fix, correct?

File: tests/test_benchmark.py [hunk 2]
```diff
@@ -229,6 +229,86 @@ def benchmark(
...
+):
+    """
+    TODO: configuration blob containing num_batches, steps_per_batch, and any other options
+    """
+
+    seed = 1234
+    dt = 1.5e-3
+    temperature = constants.DEFAULT_TEMP
+    friction = 1.0
+    seconds_per_day = 86400
```
Comment (maxentile) [inline tests/test_benchmark.py:253]: nit: style preference for `24 * 60 * 60`

File: tests/test_benchmark.py [hunk 3]
```diff
@@ -229,6 +229,86 @@ def benchmark(
...
+    # run once before timer starts
+    ctxt.multiple_steps_local(steps_per_batch, ligand_idxs, burn_in=0)
+
+    start = time.time()
+
+    for batch in range(num_batches):
+
+        # time the current batch
+        batch_start = time.time()
+        _, _ = ctxt.multiple_steps_local(steps_per_batch, ligand_idxs, burn_in=0)
```
Comment (maxentile) [inline tests/test_benchmark.py:288]: clarification: should a different random seed be set each batch?

**Review by maxentile — APPROVED** [review]
Good catch + other improvements!

File: timemachine/cpp/src/context.cu (hunk 1 above)
Comment (badisa) [reply timemachine/cpp/src/context.cu:343]: Correct.

`local_bps` contains the restraint in the vector, `bps_` does not.

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 4]
```diff
@@ -294,7 +294,8 @@ void declare_context(py::module &m) {
             The flat bottom restraint K value to use for selection and restraint of atoms within the inner shell.
 
         seed: int
-            The seed that is used to randomly select a particle to freeze.
+            The seed that is used to randomly select a particle to freeze and for the probabilistic selection of
+            free particles.
```
[the remark this answers is not in the document]
Comment (badisa) [reply timemachine/cpp/src/wrap_kernels.cpp:298]: Yes, will add

[the remark this answers is not in the document]
Comment (badisa) [reply timemachine/cpp/src/wrap_kernels.cpp]: Clarified in [f7189a7](https://github.com/proteneer/timemachine/pull/968/commits/f7189a7fe295ec7d4a67a8cefca7247dfe456179)

File: tests/test_benchmark.py (hunk 2 above)
Comment (badisa) [reply tests/test_benchmark.py]: Done in [7c8d6f1](https://github.com/proteneer/timemachine/pull/968/commits/7c8d6f161b6b3545df69eff2c85a3d78ce4bb982)

File: tests/test_benchmark.py (hunk 3 above)
Comment (badisa) [reply tests/test_benchmark.py]: Fixed in [7c8d6f1](https://github.com/proteneer/timemachine/pull/968/commits/7c8d6f161b6b3545df69eff2c85a3d78ce4bb982) and [0088560](https://github.com/proteneer/timemachine/pull/968/commits/00885600ca54445690a8abf5bf76e6f144531f33)

File: tests/test_local_move.py [hunk 5]
```diff
@@ -222,8 +230,12 @@ def test_local_md_particle_density():
     for p, bp in zip(sys_params, unbound_potentials):
         bps.append(bp.bind(p).bound_impl(np.float32))
 
-    def observable(x):
-        ixns = nblist.get_nblist(coords, box, cutoff)
+    def observable(pair):
```
Comment (mcwitt) [inline tests/test_local_move.py:233]: nit: a more descriptive name and/or comment seems useful here (took me a minute to understand what the observable was)

File: timemachine/cpp/src/context.cu [hunk 6]
```diff
@@ -332,12 +332,15 @@ std::array<std::vector<double>, 2> Context::multiple_steps_local(
         gpuErrchk(cudaStreamSynchronize(stream));
 
         // Set the nonbonded potential to compute forces of inner+outer shell.
-        set_nonbonded_potential_idxs(nonbonded_potential, p_num_selected.data[0], d_row_idxs.data, stream);
+        // Commented out as if a particle moves away from the ligand, it may have interactions with
+        // particles that aren't included in frozen
+        // set_nonbonded_potential_idxs(nonbonded_potential, p_num_selected.data[0], d_row_idxs.data, stream);
```
Comment (mcwitt) [inline timemachine/cpp/src/context.cu:337]: curiosity: was this causing issues in practice? Might be worth commenting on whether this is a correctness issue and what a fix could look like

**Review by mcwitt — APPROVED** [review]

File: tests/test_local_move.py [hunk 7]
```diff
@@ -58,15 +64,14 @@ def expect_no_drift(x0, move_fxn, observable_fxn, n_local_resampling_iterations=
...
-    # TODO: don't hard-code T, thresholds, etc.
-    T = 10
-    assert n_local_resampling_iterations > 2 * T
-    avg_at_start = np.mean(expected_selection_fraction_traj[:T])
-    avg_at_end = np.mean(expected_selection_fraction_traj[-T:])
+    avg_at_start = np.mean(expected_selection_fraction_traj[:n_samples])
+    avg_at_end = np.mean(expected_selection_fraction_traj[-n_samples:])
+    if avg_at_start == avg_at_end:
+        if np.all(expected_selection_fraction_traj == avg_at_end):
+            print("WARNING: All values identical, test may be invalid")
```
[the remark this answers is not in the document]
Comment (maxentile) [reply tests/test_local_move.py:71]: (some of the secondary assertions from the ideal gas test may be applicable here -- see also this discussion https://github.com/proteneer/timemachine/pull/785#discussion_r937865100 )

File: timemachine/cpp/src/context.cu (hunk 6 above)
Comment (badisa) [reply timemachine/cpp/src/context.cu:337]: I don't have a sense for whether this is happening in practice. Its simply that it could happen and seems safer to avoid the case entirely.

The fix is effectively to move to the Ixn group, as it will be the free vs everything else, so there will be no way for a particle to be non-interacting with another particle in the system.

File: tests/test_local_move.py [hunk 8]
```diff
@@ -46,8 +47,13 @@ def hmc_move(x0):
     return hmc_move
 
 
-def expect_no_drift(x0, move_fxn, observable_fxn, n_local_resampling_iterations=100):
-    traj = [jnp.array(x0)]
+def expect_no_drift(
```
[the remark this answers is not in the document]
Comment (badisa) [reply tests/test_local_move.py]: Changed in [2e99395](https://github.com/proteneer/timemachine/pull/968/commits/2e993956c84c75cbaaa6a508c1af8aa2dba93860)

File: tests/test_local_move.py [hunk 9]
```diff
@@ -58,15 +64,14 @@ def expect_no_drift(x0, move_fxn, observable_fxn, n_local_resampling_iterations=
...
+    avg_at_end = np.mean(expected_selection_fraction_traj[-n_samples:])
+    if avg_at_start == avg_at_end:
+        if np.all(expected_selection_fraction_traj == avg_at_end):
+            print("WARNING: All values identical, test may be invalid")
 
-    ratio = avg_at_end / avg_at_start
-    deviated_by_50percent_or_more = ratio <= 0.5 or ratio >= 1.5
-    if deviated_by_50percent_or_more:
+    percent_diff = np.abs(((avg_at_start - avg_at_end) / avg_at_start))
+    if percent_diff > threshold:
```
[the remark this answers is not in the document]
Comment (badisa) [reply tests/test_local_move.py:73]: Changed in [2e99395](https://github.com/proteneer/timemachine/pull/968/commits/2e993956c84c75cbaaa6a508c1af8aa2dba93860)

File: tests/test_local_move.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (badisa) [reply tests/test_local_move.py]: Changed in [2e99395](https://github.com/proteneer/timemachine/pull/968/commits/2e993956c84c75cbaaa6a508c1af8aa2dba93860)

File: tests/test_local_move.py (hunk 5 above)
Comment (badisa) [reply tests/test_local_move.py]: Changed in [04d13a6](https://github.com/proteneer/timemachine/pull/968/commits/04d13a6dd6e434582e7c327287c282bd3d9fe7fb)
